sia/host/contracts.go

```diff
@@ -0,0 +1,214 @@
+		err = errors.New("valid proof payout is too low")
+		return
+	}
+	// Output for failed proofs needs to be the 0 address.
+	emptyAddress := consensus.CoinAddress{}
+	if t.FileContracts[0].MissedProofAddress != emptyAddress {
+		err = errors.New("burn payout needs to go to the empty address")
+		return
+	}
+	// Verify that output for failed proofs matches burn.
+	maxBurn := h.announcement.Burn * consensus.Currency(fileSize) * consensus.Currency(t.FileContracts[0].ChallengeWindow)
+	if t.FileContracts[0].MissedProofPayout > maxBurn {
+		err = errors.New("burn payout is too high for a missed proof.")
+		return
+	}
+	// Verify that the contract fund covers the payout and burn for the whole
+	// duration.
+	requiredFund := (h.announcement.Burn + h.announcement.Price) * consensus.Currency(fileSize) * consensus.Currency(contractDuration)
+	if t.FileContracts[0].ContractFund < requiredFund {
+		err = errors.New("ContractFund does not cover the entire duration of the contract.")
+		return
+	}
+
+	// Add enough funds to the transaction to cover the penalty half of the
+	// agreement.
+	penalty := h.announcement.Burn * consensus.Currency(fileSize) * consensus.Currency(contractDuration)
+	id, err := h.wallet.RegisterTransaction(t)
+	if err != nil {
+		return
+	}
+	err = h.wallet.FundTransaction(id, penalty)
+	if err != nil {
+		// TODO: This leaks that the host is out of money.
+		return
+	}
+	updatedTransaction, err = h.wallet.SignTransaction(id, true)
+
+	return
+}
+
+// NegotiateContract is an RPC that negotiates a file contract. If the
+// negotiation is successful, the file is downloaded and the host begins
+// submitting proofs of storage.
+//
+// Care is taken not to have any locks in place when network communication is
+// happening, nor while any intensive file operations are happening. For this
+// reason, all of the locking in this function is done manually. Edit with
+// caution, review with caution.
+//
+// TODO: Split some of this logic into helper functions, which in particular
+// will make mutex management easier.
+func (h *Host) NegotiateContract(conn net.Conn) (err error) {
+	// Read the transaction from the connection.
+	var t consensus.Transaction
+	if err = encoding.ReadObject(conn, &t, maxContractLen); err != nil {
+		return
+	}
+
+	// Check that the contained FileContract fits host criteria for taking
+	// files, replying with the error if there's a problem.
+	h.lock()
+	t, err = h.considerContract(t)
+	h.unlock()
+	if err != nil {
+		_, err = encoding.WriteObject(conn, err.Error())
+		return
+	}
+	_, err = encoding.WriteObject(conn, components.AcceptContractResponse)
+	if err != nil {
+		return
+	}
+
+	// Create file.
+	h.lock()
+	filename := h.hostDir + h.getFilename()
+	h.unlock()
+	file, err := os.Create(filename)
+	if err != nil {
+		return
+	}
+	defer file.Close()
+
+	// If there's an error upon return, delete the file that's been created.
+	defer func() {
+		if err != nil {
+			os.Remove(filename)
+		}
```

Comment:
That would cause a race condition, because the lock is only in place the first time we get the name. If we were to defer a `h.fileCounter--` we'd need to keep the lock in place for the full duration of the function.

Since we're breaking mutex convention, I like making as few reads and writes as possible, because mutex mistakes are really easy to make.